crates/bevy_ecs/src/query/access.rs

```diff
@@ -54,76 +62,115 @@ impl<T: SparseSetIndex> Access<T> {
     pub fn has_write(&self, index: T) -> bool {
-        self.writes.contains(index.sparse_set_index())
+        if self.writes_all {
+            true
+        } else {
+            self.writes.contains(index.sparse_set_index())
+        }
     }
 
-    /// Sets this `Access` to having read access for all indices.
-    pub fn read_all(&mut self) {
+    /// Sets this as having access to all indexed elements.
+    pub(crate) fn read_all(&mut self) {
         self.reads_all = true;
     }
 
-    /// Returns true if this `Access` has read access to all indices.
-    pub fn reads_all(&self) -> bool {
-        self.reads_all
+    /// Sets this as having exclusive access to all indexed elements.
+    pub(crate) fn write_all(&mut self) {
+        self.reads_all = true;
+        self.writes_all = true;
     }
 
-    /// Clears all recorded accesses.
+    /// Removes all accesses.
     pub fn clear(&mut self) {
         self.reads_all = false;
+        self.writes_all = false;
         self.reads_and_writes.clear();
         self.writes.clear();
     }
 
-    /// Extends this `Access` with another, copying all accesses of `other` into this.
+    /// Adds all access from `other`.
     pub fn extend(&mut self, other: &Access<T>) {
         self.reads_all = self.reads_all || other.reads_all;
+        self.writes_all = self.writes_all || other.writes_all;
         self.reads_and_writes.union_with(&other.reads_and_writes);
         self.writes.union_with(&other.writes);
     }
 
-    /// Returns true if this `Access` is compatible with `other`.
+    /// Returns `true` if this and `other` can be active at the same time.
     ///
-    /// Two `Access` instances are incompatible with each other if one `Access` has a write for
-    /// which the other also has a write or a read.
+    /// `Access` instances are incompatible if one or both wants exclusive access to an element they have in common
+    /// (data race).
     pub fn is_compatible(&self, other: &Access<T>) -> bool {
-        if self.reads_all {
-            0 == other.writes.count_ones(..)
+        if self.writes_all {
+            // self has `&mut World`, other could have anything
+            other.reads_and_writes.count_ones(..) == 0
+        } else if other.writes_all {
+            // other has `&mut World`, self could have anything except `&mut World`
+            self.reads_and_writes.count_ones(..) == 0
+        } else if self.reads_all {
+            // self has `&World`, other could have anything except `&mut World`
+            other.writes.count_ones(..) == 0
         } else if other.reads_all {
-            0 == self.writes.count_ones(..)
+            // other has `&World`, self has neither `&World` or `&mut World`
+            self.writes.count_ones(..) == 0
         } else {
+            // neither system has `&World` or `&mut World`
             self.writes.is_disjoint(&other.reads_and_writes)
                 && self.reads_and_writes.is_disjoint(&other.writes)
         }
     }
```

Comment:
These comparisons are "correct" for `&World` and `&mut World`. If we should keep public setters for the `reads_all` and `writes_all` fields, somebody give code that properly generalizes.

Reply:
```rust
    pub fn is_compatible(&self, other: &Access<T>) -> bool {
         // Only systems with no access are compatible with systems that operate on &mut World
         if self.writes_all & other.reads_and_writes.count_ones > 0 {
            return false;
         }

         if other.writes_all & self.reads_and_writes.count_ones > 0 {
            return false;
         }

         // Only systems that do not write data are compatible with systems that operate on &World
         if self.reads_all & other.writes.count_ones > 0 {
            return false;
         }
         
         if other.reads_all & self.writes.count_ones > 0 {
            return false;
         }

        // Insert standard logic here
        ... 
    }
```

Here's a refactored version of the logic that I think is a fair bit more readable by using early returns.